internal/controllers/machinedeployment/machinedeployment_controller.go

```diff
@@ -332,24 +333,65 @@ func (r *Reconciler) reconcile(ctx context.Context, s *scope) error {
 	log := ctrl.LoggerFrom(ctx)
-	allMSs := append(p.oldMSs, p.newMS)
 
+	// Note: newMS goes first so in the logs we will have first create/scale up newMS, then scale down oldMSs
+	allMSs := append([]*clusterv1.MachineSet{p.newMS}, p.oldMSs...)
+
+	// Get all the diff introduced by the rollout planner.
+	// Note: collect all the diff first, so for each change we can add an overview of all the MachineSets
+	// in the MachineDeployment, because those info are required to understand why a change happened.
+	machineSetsDiff := map[string]machineSetDiff{}
+	machineSetsSummary := map[string]string{}
+	for _, ms := range allMSs {
+		// Update spec.Replicas in the MachineSet.
+		if scaleIntent, ok := p.scaleIntents[ms.Name]; ok {
+			ms.Spec.Replicas = ptr.To(scaleIntent)
+		}
+
+		diff := p.getMachineSetDiff(ms)
+		machineSetsDiff[ms.Name] = diff
+		machineSetsSummary[ms.Name] = diff.Summary
+	}
+
+	// Apply changes to MachineSets.
 	for _, ms := range allMSs {
 		log = log.WithValues("MachineSet", klog.KObj(ms))
 		ctx = ctrl.LoggerInto(ctx, log)
 
-		if scaleIntent, ok := p.scaleIntents[ms.Name]; ok {
-			ms.Spec.Replicas = &scaleIntent
+		// Retrieve the diff for the MachineSet.
+		// Note: no need to check for diff doesn't exist, all the diff have been computed right above.
+		diff := machineSetsDiff[ms.Name]
+
+		// Add to the log kv pairs providing context and details about changes in this MachineSet (reason, diff),
+		// and also add the overview of all the MachineSets computed above.
+		// Note: Those values should not be added to the context to prevent propagation to other func.
+		statusToLogKeyAndValues := []any{
+			"reason", diff.Reason,
+			"diff", diff.OtherChanges,
+			"machinesets", machineSetsSummary,
 		}
+		if len(p.acknowledgedMachineNames) > 0 {
+			statusToLogKeyAndValues = append(statusToLogKeyAndValues, "acknowledgedMachines", sortAndJoin(p.acknowledgedMachineNames))
+		}
+		if len(p.updatingMachineNames) > 0 {
+			statusToLogKeyAndValues = append(statusToLogKeyAndValues, "updatingMachines", sortAndJoin(p.updatingMachineNames))
+		}
+		log := log.WithValues(statusToLogKeyAndValues...)
 
-		if ms.GetUID() == "" {
+		if diff.OriginalMS == nil {
 			// Create the MachineSet.
 			if err := ssa.Patch(ctx, r.Client, machineDeploymentManagerName, ms); err != nil {
 				r.recorder.Eventf(p.md, corev1.EventTypeWarning, "FailedCreate", "Failed to create MachineSet %s: %v", klog.KObj(ms), err)
 				return errors.Wrapf(err, "failed to create MachineSet %s", klog.KObj(ms))
 			}
-			log.Info(fmt.Sprintf("MachineSet created (%s)", p.createReason))
-			r.recorder.Eventf(p.md, corev1.EventTypeNormal, "SuccessfulCreate", "Created MachineSet %s with %d replicas", klog.KObj(ms), ptr.Deref(ms.Spec.Replicas, 0))
+			if len(p.oldMSs) > 0 {
+				log.Info(fmt.Sprintf("MachineSets need rollout: %s", strings.Join(machineSetNames(p.oldMSs), ", ")), "reason", p.createReason)
```

Comment:
This adds another reason k/v pair

(I don't remember if on that level klog overwrites or duplicates the k/v pair, but both is not desired)